src/core/blueprint/wrapping_type.rs

```diff
@@ -173,3 +175,235 @@ impl From<String> for Type {
+    /// That's a wide merge i.e. the result wides the input definitions
+    /// the way it could be handled by both self and other sources
+    fn federated_merge_union(self, other: Self) -> Valid<Self, String> {
+        #[inline]
+        fn non_null_merge(non_null: bool, other_non_null: bool) -> bool {
+            non_null && other_non_null
+        }
+
+        self.merge(other, non_null_merge)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    mod federated_merge_wide {
+        use super::*;
+
+        #[test]
+        fn test_equal() {
+            let a = Type::Named { name: "String".to_owned(), non_null: false };
+            let b = Type::Named { name: "String".to_owned(), non_null: false };
+
+            assert_eq!(
+                a.federated_merge_union(b),
+                Valid::succeed(Type::Named { name: "String".to_owned(), non_null: false })
+            );
+
+            let a = Type::List {
+                of_type: Box::new(Type::Named { name: "Int".to_owned(), non_null: false }),
+                non_null: true,
+            };
+            let b = Type::List {
+                of_type: Box::new(Type::Named { name: "Int".to_owned(), non_null: false }),
+                non_null: true,
+            };
+
+            assert_eq!(
+                a.federated_merge_union(b),
+                Valid::succeed(Type::List {
+                    of_type: Box::new(Type::Named { name: "Int".to_owned(), non_null: false }),
+                    non_null: true,
+                })
+            );
+        }
+
+        #[test]
+        fn test_different_non_null() {
+            let a = Type::Named { name: "String".to_owned(), non_null: false };
+            let b = Type::Named { name: "String".to_owned(), non_null: true };
+
+            assert_eq!(
+                a.federated_merge_union(b),
+                Valid::succeed(Type::Named { name: "String".to_owned(), non_null: false })
+            );
+
+            let a = Type::List {
+                of_type: Box::new(Type::Named { name: "Int".to_owned(), non_null: false }),
+                non_null: false,
+            };
+            let b = Type::List {
+                of_type: Box::new(Type::Named { name: "Int".to_owned(), non_null: true }),
+                non_null: true,
+            };
+
+            assert_eq!(
+                a.federated_merge_union(b),
+                Valid::succeed(Type::List {
+                    of_type: Box::new(Type::Named { name: "Int".to_owned(), non_null: false }),
+                    non_null: false,
+                })
+            );
+        }
+
+        #[test]
+        fn test_different_types() {
+            let a = Type::Named { name: "String".to_owned(), non_null: false };
+            let b = Type::Named { name: "Int".to_owned(), non_null: false };
+
+            assert_eq!(
+                a.federated_merge_union(b),
+                Valid::fail("Type mismatch: expected `String`, got `Int`".to_owned())
```

Comment:
Field name could also be provided.

Reply:
what do you mean by that? this code checks the merging of the type of the field, it's a part of field merging but without the field name